### PR TITLE
#718 #730: Add Gray8 and Gray16 Pixel Formats and clean up IPixel

### PR DESCRIPTION
**Pull request by jongleur1983**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Implements Pixel formats Gray8 and Gray16.
Where I knew how to do it I provided optimized conversion methods, where not I used the "fallback" strategy to convert from and to Vector4, as it's done for many other implementations as well.

Unit tests are based on some of the other Pixel formats tests. Not sure this is seen as sufficient though, please tell me if not.

I guess I have to squash commits again, but for review the individual commits may help so I'm happy to do so after a first review.

### CODE REVIEW

File: src/ImageSharp/PixelFormats/Argb32.cs [hunk 1]
```diff
@@ -289,6 +290,35 @@ public void ToBgra32(ref Bgra32 dest)
...
+        {
+            this.A = 255;
+            this.R = source.PackedValue;
+            this.G = source.PackedValue;
+            this.B = source.PackedValue;
+        }
+
+        /// <inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void ToGray8(ref Gray8 dest) => dest.PackFromScaledVector4(this.ToScaledVector4());
```
Comment (JimBobSquarePants) [inline src/ImageSharp/PixelFormats/Argb32.cs]: This can be optimized.

File: src/ImageSharp/PixelFormats/Gray8.cs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        public Vector4 ToVector4()
+        {
+            return new Vector4(this.PackedValue, this.PackedValue, this.PackedValue, 1.0f);
+        }
+
+        /// <inheritdoc />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void PackFromRgba32(Rgba32 source)
+        {
+            this.PackedValue = Pack(source.R, source.G, source.B);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/PixelFormats/Gray8.cs]: This and others is incorrect. Will lead to a packed value of 255.

File: src/ImageSharp/PixelFormats/Gray8.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        /// <inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void PackFromRgb48(Rgb48 source) =>
+            this.PackedValue = Pack(source.R, source.G, source.B);
+
+        /// <inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void ToRgb48(ref Rgb48 dest)
+        {
+            ushort gray = (ushort)(this.PackedValue * 255);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/PixelFormats/Gray8.cs]: `* 257`

File: src/ImageSharp/PixelFormats/Gray8.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+        /// <summary>
+        /// Packs a <see cref="float"/> into a byte.
+        /// </summary>
+        /// <param name="r">Red value of the color to pack.</param>
+        /// <param name="g">Green value of the color to pack.</param>
+        /// <param name="b">Blue value of the color to pack.</param>
+        /// <returns>The <see cref="byte"/> containing the packed value.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static byte Pack(float r, float g, float b)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/PixelFormats/Gray8.cs]: This only works with `Vector4` but is being used for `Rgba32` etc.

**Comment by JimBobSquarePants** [comment]
I'm gonna do #730 at the same time otherwise we're writing code we'll immediately delete.

**Comment by JimBobSquarePants** [comment]
@jongleur1983 Ok. So I've fixed up all the pixels formats and commented out all the old tests. 

I need to do two things now to complete this PR:

1. Add a new suite of tests to cover IPixel functionality
2. Add optimized overloads for the PixelOperations methods.

**Comment by JimBobSquarePants** [comment]
Got four tests failing due to `ToString` change in `Rgba32` as the reference image path depended on original name. We should update that. Perhaps hex?

https://ci.appveyor.com/project/six-labors/imagesharp/builds/19386791/job/x9ru3njufhn9ec6j/tests

**Comment by antonfirsov** [comment]
Yeah hex is better for this purpose

**Comment by JimBobSquarePants** [comment]
Ready for review.

Code coverage is a bit less than I would like but I'm done now. It's been 10 days writing this.

File: src/ImageSharp/PixelFormats/Generated/PixelOperations{TPixel}.Generated.cs [hunk 5]
```diff
@@ -10,320 +10,379 @@ namespace SixLabors.ImageSharp.PixelFormats
...
+            ref Argb32 destBaseRef = ref MemoryMarshal.GetReference(destPixels);
 
             for (int i = 0; i < count; i++)
             {
                 ref TPixel sp = ref Unsafe.Add(ref sourceBaseRef, i);
-                ref Rgba64 dp = ref Unsafe.Add(ref destBaseRef, i);
-                sp.ToRgba64(ref dp);
+                ref Argb32 dp = ref Unsafe.Add(ref destBaseRef, i);
+
+                dp.PackFromScaledVector4(sp.ToScaledVector4());
```
Comment (antonfirsov) [inline src/ImageSharp/PixelFormats/Generated/PixelOperations{TPixel}.Generated.cs]: This is again a significant performance regression. Many pixel formats had efficient implementations for the conversion of basic formats like (`Rgba32` <-> `Bgra32` <-> `Argb32` <-> `Rgb24` etc, implemented with byte-shuffling.

We should not replace those with expensive internal `Vector4` conversions!

**Comment by antonfirsov** [comment]
Ok, actually this *might* be good, I haven't noticed the new optimized bulk implementations in my review!

File: src/ImageSharp/PixelFormats/IPixel.cs [hunk 6]
```diff
@@ -48,83 +42,77 @@ public interface IPixel
...
+        void PackFromRgba32(Rgba32 source);
 
         /// <summary>
-        /// Converts the pixel to <see cref="Argb32"/> format.
+        /// Expands the packed representation into an <see cref="Rgba32"/>.
         /// </summary>
-        /// <param name="dest">The destination pixel to write to</param>
-        void ToArgb32(ref Argb32 dest);
+        /// <returns>The <see cref="Rgba32"/>.</returns>
+        Rgba32 ToRgba32();
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/PixelFormats/IPixel.cs]: The good thing is that this method is rarely used. We can probably address this later together with the review on `ToVector4()` / `ToScaledVector4()`.

**Comment by JimBobSquarePants** [comment]
@antonfirsov Yeah, all the `IPixel` conversions have optimized bulk implementations so there should be no performance loss. It should actually be faster in some cases than before.

File: src/ImageSharp/PixelFormats/IPixel.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/PixelFormats/IPixel.cs]: I don't follow. Are you saying that `ToRgba()` is 9x slower than `ToRgba(ref Rgba)`? I would have thought the cost of the pointer would outweigh any benefit especially in 64bit mode.

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/PixelFormats/IPixel.cs]: Depends on the use case.
The case in the benchmark is basically the `Span<TPixel>` -> `Span<Rgba32>` conversion. Writing right into the reference of the destination buffer removes one extra copy, compared to the case when you are assigning the destiniation buffer item through a return value.

I suspect, there are probably some other unnecessary operations JIT fails to optimize.

We should also benchmark an other case:
When the result of `ToRgba()` is being copied into a temporary variable for a small modification before writing it to the destination buffer. This is what we do in the PNG codec.

We should implement these benchmarks for `ToVector4()` as well. Gonna work on these points tonight!

**Comment by JimBobSquarePants** [comment]
@antonfirsov Those results are extraordinary, I'm very surprised!

Let's do it here now. A new PR will be another very large one again probably too large to digest. We know where we are here and can review knowing exactly what we are looking at having recently worked on the files.

**Comment by antonfirsov** [comment]
One possible explanation is that JIT is not capable for [Return value optimization](https://en.wikipedia.org/wiki/Copy_elision#Return_value_optimization) in most cases.

Do you mean also finishing up the `Vector4` stuff here? Should/can we split the work somehow?

**Comment by JimBobSquarePants** [comment]
I was tempted to do it all here but for sanities sake lets do `Rgba32` here and immediately raise an issue for next priority for `Vector4`. We don't want to be blocking any open PR's

**Comment by antonfirsov** [comment]
@JimBobSquarePants my feeling is that it would be much simpler to merge this as is, right after #742 and carry out all my proposed API changes together in a new PR.

**Comment by JimBobSquarePants** [comment]
@antonfirsov works for me! 👍

**Comment by antonfirsov** [comment]
@JimBobSquarePants any idea how to ping the CLA assistant? Pressing "Recheck PR-s" on it's UI did not help.

**Comment by JimBobSquarePants** [comment]
@antonfirsov Github's having all sorts of issues just now.

**Pull request merged**